neuraxle/checkpoints.py

```diff
@@ -19,185 +19,558 @@
+        'Checkpoint', DataContainer]:
+        """
+        Saves step, and data checkpointers for the FIT execution mode.
+
+        :param data_container: data container for creating the data checkpoint
+        :param context: context for creating the step checkpoint
+        :return: saved data container
+        :rtype: DataContainer
+        """
+        self.save_checkpoint(data_container, context)
         return self, data_container
 
     def handle_transform(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
-        data_container = self._handle_any(context, data_container)
-        return data_container
+        """
+        Saves step, and data checkpointers for the TRANSORM execution mode.
 
-    def handle_fit_transform(self, data_container: DataContainer, context: ExecutionContext) -> ('BaseStep', DataContainer):
-        data_container = self._handle_any(context, data_container)
-        return self, data_container
+        :param data_container: data container for creating the data checkpoint
+        :param context: context for creating the step checkpoint
+        :return: saved data container
+        :rtype: DataContainer
+        """
+        return self.save_checkpoint(data_container, context)
 
-    def _handle_any(self, context, data_container):
-        self.set_checkpoint_path(context.get_path())
-        data_container: DataContainer = self.save_checkpoint(data_container)
+    def handle_fit_transform(self, data_container: DataContainer, context: ExecutionContext) -> Tuple[
+        'Checkpoint', DataContainer]:
+        """
+        Saves step, and data checkpointers for the FIT_TRANSORM execution mode.
 
-        self.save_checkpoint(data_container)
-        context.save_all_unsaved()
+        :param data_container: data container for creating the data checkpoint
+        :param context: context for creating the step checkpoint
+        :return: saved data container
+        :rtype: DataContainer
+        """
+        return self, self.save_checkpoint(data_container, context)
+
+    def save_checkpoint(self, data_container: DataContainer, context: ExecutionContext):
+        """
+        Saves step, and data checkpointers for the current execution mode.
+
+        :param data_container: data container for creating the data checkpoint
+        :param context: context for creating the step checkpoint
+        :return: saved data container
+        :rtype: DataContainer
+        """
+        if self.step_checkpointer is not None and self.step_checkpointer.is_for_execution_mode(
+                context.get_execution_mode()):
+            self.step_checkpointer.save_checkpoint(data_container, context)
+
+        for checkpointer in self.data_checkpointers:
+            if checkpointer.is_for_execution_mode(context.get_execution_mode()):
+                checkpointer.save_checkpoint(data_container, context)
```

Comment:
This code is uselessly segregated. I'd finally ask to merge self.step_checkpointer and self.data_checkpointers into one variable. 
```
for checkpointer in self.all_checkpointers: 
    checkpointer.save_checkpoint(data_container, context)
```
And don't even do the `if checkpointer.is_for_execution_mode(context.get_execution_mode()):`, instead do this if from inside each checkpointers inside the method `checkpointer.save_checkpoint(data_container, context)`. I think this would less break the Law of Demeter.